src/main/scala/sangria/execution/QueryReducerExecutor.scala

```diff
@@ -0,0 +1,179 @@
+import sangria.schema._
+import sangria.util.tag.@@
+import sangria.validation.QueryValidator
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+import scala.util.{Failure, Success}
+
+object QueryReducerExecutor {
+
+  def reduceQueryWithoutVariables[Ctx, Root](
+    schema: Schema[Ctx, Root],
+    queryAst: ast.Document,
+    userContext: Ctx,
+    root: Root,
+    queryReducers: List[QueryReducer[Ctx, _]],
+    operationName: Option[String] = None,
+    queryValidator: QueryValidator = QueryValidator.default,
+    exceptionHandler: ExceptionHandler = ExceptionHandler.empty,
+    deprecationTracker: DeprecationTracker = DeprecationTracker.empty,
+    middleware: List[Middleware[Ctx]] = Nil
+    )(implicit executionContext: ExecutionContext): Future[(Ctx, TimeMeasurement)] = {
+    val violations = queryValidator.validateQuery(schema, queryAst)
+
+    if (violations.nonEmpty)
+      Future.failed(ValidationError(violations, exceptionHandler))
+    else {
+      val scalarMiddleware = Middleware.composeFromScalarMiddleware(middleware, userContext)
+      val valueCollector = new UnknownVariablesValueCollector[Ctx](schema, queryAst.sourceMapper, deprecationTracker, userContext, exceptionHandler, scalarMiddleware)
+
+      val executionResult = for {
+        operation ← Executor.getOperation(exceptionHandler,queryAst, operationName)
+        fieldCollector = new FieldCollector[Ctx, Root](schema, queryAst, Map.empty, queryAst.sourceMapper, valueCollector, exceptionHandler)
+        tpe ← Executor.getOperationRootType(schema, exceptionHandler, operation, queryAst.sourceMapper)
+        fields ← fieldCollector.collectFields(ExecutionPath.empty, tpe, Vector(operation))
+      } yield QueryReducerExecutor.reduceQuery(schema, queryReducers, exceptionHandler, fieldCollector, valueCollector, tpe, fields, userContext)
+
+      executionResult match {
+        case Success(future) ⇒ future
+        case Failure(error) ⇒ Future.failed(error)
+      }
+    }
+  }
+
+  /** Returns either new Ctx or future of it (with time measurement) */
+  def reduceQuery[Ctx, Root, Val](
+    schema: Schema[Ctx, Root],
+    queryReducers: List[QueryReducer[Ctx, _]],
+    exceptionHandler: ExceptionHandler,
+    fieldCollector: FieldCollector[Ctx, Root],
+    valueCollector: ValueCollector[Ctx, _],
+    rootTpe: ObjectType[Ctx, Root],
+    fields: CollectedFields,
+    userContext: Ctx)(implicit executionContext: ExecutionContext): Future[(Ctx, TimeMeasurement)] =
+    if (queryReducers.nonEmpty) {
+      val sw = StopWatch.start()
+      reduceQueryUnsafe(schema, fieldCollector, valueCollector, rootTpe, fields, queryReducers.toVector, userContext)
+        .map(_ → sw.stop)
+        .recover { case error: Throwable ⇒ throw QueryReducingError(error, exceptionHandler) }
+    } else Future.successful(userContext → TimeMeasurement.empty)
+
+  private def reduceQueryUnsafe[Ctx, Val](
```

Comment:
This probably was not the easiest part of the codebase to read through :) It is quite mutable and untyped. Not very proud of this code, but this approach helps to reduce the object allocation. The main idea was to reduce performance impact as much as possible.